### PR TITLE
Blog app - integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by anshuman7negi**
**In this branch we worked on the following task=>**

N+1 problem is solved when fetching all posts and their comments for a user.

Added test cases for:

- User index page:
  - I can see the username of all other users.
  - I can see the profile picture for each user.
  - I can see the number of posts each user has written.
  - When I click on a user, I am redirected to that user's show page. 
  
- user show page:
  - I can see the user's profile picture.
  - I can see the user's username.
  - I can see the number of posts the user has written.
  - I can see the user's bio.
  - I can see the user's first 3 posts.
  - I can see a button that lets me view all of a user's posts.
  - When I click a user's post, it redirects me to that post's show page.
  - When I click to see all posts, it redirects me to the user's post's index page.
  
- User post index page:
  - I can see the user's profile picture.
  - I can see the user's username.
  - I can see the number of posts the user has written.
  - I can see a post's title.
  - I can see some of the post's body.
  - I can see the first comments on a post.
  - I can see how many comments a post has.
  - I can see how many likes a post has.
  - I can see a section for pagination if there are more posts than fit on the view.
  - When I click on a post, it redirects me to that post's show page.

- Post show page:
  - I can see the post's title.
  - I can see who wrote the post.
  - I can see how many comments it has.
  - I can see how many likes it has.
  - I can see the post body.
  - I can see the username of each commenter.
  - I can see the comment each commentor left.

 **screenshot  before N + 1 problem solved**
image

 **screenshot  after N + 1 problem solved**
image

### CODE REVIEW

File: spec/integration/user_show_spec.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  scenario 'displays the user profile information' do
+    visit user_path(@first_user)
+    expect(page).to have_selector("img[src*='https://example.com/user1.jpg']")
+    expect(page).to have_content('Tom')
+    expect(page).to have_content('Number of posts: 4')
+    expect(page).to have_content('Teacher from Mexico')
+
+    [@post4, @post3, @post2].each do |post|
+      expect(page).to have_content(post.text)
+    end
```
Comment (mckent05) [inline spec/integration/user_show_spec.rb]: - You have done well so far, but to test if we can see the users first 3 posts, please use the `recent_posts` method in the user model. Please fix this 🙏

File: spec/integration/posts_index_spec.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+                       likes_counter: 5)
+
+    Comment.create(post:, user: @user, text: 'Great post!')
+    Comment.create(post:, user: @user, text: 'I love it!')
+
+    visit user_posts_path(@user)
+
+    expect(page).to have_content('This is the first post.')
+    expect(page).to have_content("Comments: #{post.comments_counter}")
+    expect(page).to have_content("Likes: #{post.likes_counter}")
```
Comment (mckent05) [inline spec/integration/posts_index_spec.rb:36]: - Good job done so far, but please add a test to show the post title can be seen. Please fix this 🙏

**Review by mckent05 — CHANGES_REQUESTED** [review]
Hi @anshuman7negi, 

# STATUS: CHANGES REQUIRED ♻️ ♻️ 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
almost there

## Highlights
- All linter checks are passing ✔️ ✔️ 
- Your PR is professional 👍 👍 
- Good job adding integration spec for your views 👍 👍 


## Required Changes ♻️
_Check the comments under the review._
**[REQUIRED] - You have done well so far, but according to the project requirements, you are required to fix the `N+1`  problems in your app. In your `post#index` action please make sure you retrieve the comments along with all posts. i.e***
```ruby
       @posts = @user.posts.includes(:comments)
```

## Optional suggestions
Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/integration/user_show_spec.rb (hunk 1 above)
Comment (anshuman7negi) [reply spec/integration/user_show_spec.rb]: done

File: spec/integration/posts_index_spec.rb (hunk 2 above)
Comment (anshuman7negi) [reply spec/integration/posts_index_spec.rb:36]: done

**Review by uwadonat — APPROVED** [review]
## Hi @anshuman7negi,

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
### Congratulations! 🎉 
- Correct workflow was used ✔️
- Well-organized and clean project ✔️
- No linter error ✔️

approv

# Status: Approved ✔️✔️✔️

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100%.
Please, remember to tag me @uwadonat in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
```

**Pull request merged**